Sources/SwiftFusion/Core/Manifold.swift

```diff
@@ -0,0 +1,110 @@
+/// A coordinate in a differentiable manifold's global coordinate system.
+///
+/// Includes a coordinate chart [1] between `Self` and `R^n`, centered at `self`.
+///
+/// [1] https://en.wikipedia.org/wiki/Topological_manifold#Coordinate_charts
+public protocol ManifoldCoordinate: Differentiable {
+  /// The local coordinate system used in the chart.
+  ///
+  /// Isomorphic to `R^n`, where `n` is the dimension of the manifold.
+  associatedtype LocalCoordinate: AdditiveArithmetic & Differentiable & VectorProtocol
```

Comment:
This is important, maybe everything will make sense if I can successfully explain why there is a `ManifoldCoordinate.LocalCoordinate` that is distinct from `ManifoldCoordinate.TangentVector`.

The high level idea behind this PR is:

1. When a user defines a struct, Swift can automatically generate a tangent space for the struct as the direct product of the struct's properties' tangent spaces. This is nice because derivatives involving the struct just work without any boilerplate.
2. For SwiftFusion's geometry, we want a different tangent space. We can achieve this by defining a fully custom `TangentVector` type (for an example, see `Pose2.swift` in the master branch as it is today). This involves an inconvenient amount of hard-to-explain boilerplate.
3. The tangent spaces in (1) and (2) are isomorphic to each other. This suggests a strategy for getting the advantages of (1) with the tangent space of (2): Do all the derivatives in the automatically-generated tangent space from (1), but map the derivatives through the isomorphism to (2) before exposing them to the user.

There are many ways to implement the strategy in (3), and this PR is one way. This PR works like this:

* `ManifoldCoordinate.TangentVector` is generated automatically (1).
* `ManifoldCoordinate.LocalCoordinate` is the user-specified tangent space (2).
* `ManifoldCoordinate.global/local` (aka `retract/localCoordinates`) determine the isomorphism between (1) and (2).
* `Manifold` is a wrapper type with `Manifold.TangentVector == ManifoldCoordinate.LocalCoordinate`, so that clients who use the `Manifold` wrapper type get the user-specified tangent space (2) without having to worry about translating between it and the automatically-generated tangent space (1).